docs/spec/ibc/queues.md

```diff
@@ -0,0 +1,160 @@
+
+A message queue is simply a set of serialized packets stored at predefined keys in a merkle store, which can produce proofs for any key. Once a packet is written it must be immutable (except for deleting when popped from the queue). That is, if a value _v_ is written to a queue, then every valid proof _M<sub>k,v,h </sub>_ must refer to the same _v_. This property is essential to safely process asynchronous messages.
+
+Every IBC implementation must provide a protected subspace of the merkle store for use by each queue that cannot be affected by other modules.
+
+### 3.2   Naming Queues
+
+As mentioned above, in order for the receiver to unambiguously interpret the merkle proofs, we need a unique, deterministic, and predictable key in the merkle store for each message in the queue. We explained how the indexes are generated to provide each message in a queue a unique key, and mentioned the need for a unique name for each queue.
+
+The queue name must be unambiguously associated with a given connection to another chain, so an observer can prove if a message was intended for chain A or chain B. In order to do so, upon registration of a connection with a remote chain, we create two queues with different names (prefixes).
+
+* _ibc:<chain id of A>:send_ - all outgoing packets destined to chain A
+* _ibc:<chain id of A>:receipt_ - the results of executing the packets received from chain A
+
+These two queues have different purposes and store messages of different types. By parsing the key of a merkle proof, a recipient can uniquely identify which queue, if any, this message belongs to. We now define _k =_ _(remote id, [send|receipt], index)_. This tuple is used to route and verify every message, before the contents of the packet are processed by the appropriate application logic.
+
+### 3.3   Message Contents
+
+Up to this point, we have focused on the semantics of the message key, and how we can produce a unique identifier for every possible message in every possible connection. The actual data written at the location has been left as an opaque blob, put by providing some structure to the messages, we can enable more functionality.
+
+We define every message in a _send queue_ to consist of a well-known type and opaque data. The IBC protocol relies on the type for routing, and lets the appropriate module process the data as it sees fit. The _receipt queue_ stores if it was an error, an optional error code, and an optional return value. We use the same index as the received message, so that the results of _A:q<sub>B.send</sub>[i]_ are stored at _B:q<sub>A.receipt</sub>[i]_. (read: the message at index _i_ in the _send_ queue for chain B as stored on chain A)
```

Comment:
Routing is never defined.

Reply:
Unclear.
Routing happens inside one blockchain to the appropriate dapp/smart contract.
It's not part of the IBC communication protocol.

Reply:
Updated IBC send packet definition to include an opaque **type** field for routing - this is only necessary to pick a handler on the destination chain, which needs to return `success`/`failure`.